packages/viz/src/lib/layer/LayerInteractivity.ts

```diff
@@ -21,122 +18,94 @@ export class LayerInteractivity {
     layer: StyledLayer,
     layerGetStyleFn: () => Style,
     layerSetStyleFn: (style: Style) => Promise<void>,
+    layerEmitFn: (type: string, event?: unknown) => void,
+    layerOnFn: EventHandler,
+    layerOffFn: EventHandler,
     hoverStyle?: Style | string,
     clickStyle?: Style | string
   ) {
-    this._props = {};
     this._layer = layer;
-    this._layerGetStyleFn = layerGetStyleFn;
-    this._layerSetStyleFn = layerSetStyleFn;
 
     this._hoverStyle = hoverStyle;
     this._clickStyle = clickStyle;
 
+    this._layerGetStyleFn = layerGetStyleFn;
+    this._layerSetStyleFn = layerSetStyleFn;
+
+    this._layerOnFn = layerOnFn;
+    this._layerOffFn = layerOffFn;
+
     if (this._clickStyle) {
-      this.on(EventType.CLICK, () => {
+      layerOnFn(EventType.CLICK, () => {
         const interactiveStyle = this._wrapInteractiveStyle();
         this._layerSetStyleFn(interactiveStyle);
       });
     }
 
     if (this._hoverStyle) {
-      this.on(EventType.HOVER, () => {
+      layerOnFn(EventType.HOVER, () => {
         const interactiveStyle = this._wrapInteractiveStyle();
         this._layerSetStyleFn(interactiveStyle);
       });
     }
 
-    this._setCursorOnHover = false;
+    this._layerEmitFn = layerEmitFn;
   }
 
-  public getProps() {
-    return this._props;
+  public onClick(info: any, event: HammerInput) {
+    this.fireOnEvent(EventType.CLICK, info, event);
   }
 
-  public setDeckInstance(deckInstance: Deck) {
-    this._deckInstance = deckInstance;
-
-    if (this._clickPopup) {
-      this._clickPopup.addTo(this._deckInstance);
-    }
-
-    if (this._hoverPopup) {
-      this._hoverPopup.addTo(this._deckInstance);
-    }
+  public onHover(info: any, event: HammerInput) {
+    this.fireOnEvent(EventType.HOVER, info, event);
   }
 
-  /**
-   * TODO Attaches an event handler function defined by the user to
-   * this layer.
-   *
-   * @param eventType - Event type
-   * @param eventHandler - Event handler defined by the user
-   */
-  public on(eventType: EventType, eventHandler?: InteractionHandler) {
-    if (!eventHandler) {
-      if (eventType === EventType.CLICK) {
-        this._props.onClick = undefined;
-
-        // if onHover is setting the cursor
-        // then remove it
-        if (this._setCursorOnHover) {
-          this._props.onHover = undefined;
-          this._setCursorOnHover = false;
-        }
-      } else if (eventType === EventType.HOVER) {
-        // if onClick is set then onHover is just
-        // setting the cursor
-        if (this._props.onClick) {
-          this._props.onHover = this._setStyleCursor.bind(this);
-          this._setCursorOnHover = true;
-        } else {
-          this._props.onHover = undefined;
-        }
-      }
-    } else {
-      const layerHandlerFn = (info: any, event: HammerInput) => {
-        const features = [];
-        const { coordinate, object } = info;
+  public fireOnEvent(eventType: EventType, info: any, event: HammerInput) {
```

Comment:
Should we use a more specific naming for this enum, like `InteractivityEventType`?